Set timeout for optimistically upgraded navigations based on current network conditions

sysrqb opened the pull request:
#### 50f926d27862352367373c172fc56426b225789e
<pre>
Set timeout for optimistically upgraded navigations based on current network conditions
<a href="https://bugs.webkit.org/show_bug.cgi?id=278504">https://bugs.webkit.org/show_bug.cgi?id=278504</a>
<a href="https://rdar.apple.com/134458774">rdar://134458774</a>

Reviewed by Alex Christensen.

This change implements a mechanism for dynamically setting the connection
timeout for requests that were optimistically upgraded. Currently, the timeout
is hard-coded, but there won&apos;t be a single hard-coded value that will be
appropriate for all networks. Therefore, in this patch I added a mechanism for
tracking the last 25 requests that used newly established TLS connections and
computes the exponentially weighed moving average (EWMA).

This patch sets a minimum (and default) timeout of 3 seconds, and the EMWA
skews the average toward more recent entries.

* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::startNetworkLoad):

Apply the appropriate timeout.

(WebKit::NetworkResourceLoader::didReceiveResponse):

Record this load, if appropriate.

* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::NetworkSession):
(WebKit::connectionTimesMovingAverage):
(WebKit::NetworkSession::recordHTTPSConnectionTiming):

Implement EWMA and compute the current average.

* Source/WebKit/NetworkProcess/NetworkSession.h:
(WebKit::NetworkSession::currentHTTPSConnectionAverageTiming const):

Canonical link: <a href="https://commits.webkit.org/283494@main">https://commits.webkit.org/283494@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b00a54a25cbc15da10925333d85925e8b9f93ed6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66419 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45794 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19040 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70452 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/17552 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53593 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17312 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53270 "Passed tests") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/11880 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69486 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42213 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57494 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33924 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38884 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15905 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60784 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15224 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72155 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10376 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14598 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60597 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10408 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57563 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60912 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8566 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2175 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10068 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41601 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42678 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43861 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42421 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkSession.h:
```diff
@@ -354,6 +357,13 @@ class NetworkSession : public WebCore::SWServerDelegate, public CanMakeCheckedPt
+    struct RecentHTTPSConnectionTiming {
+        std::array<Seconds, 25> recentConnectionTimings;
+        std::array<Seconds, 25>::iterator front;
+        bool isArrayFull;
```
Let's give these initial values of false and 0 to prevent using uninitialized memory.

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkSession.h:
```diff
@@ -354,6 +357,13 @@ class NetworkSession : public WebCore::SWServerDelegate, public CanMakeCheckedPt
 
+    struct RecentHTTPSConnectionTiming {
+        std::array<Seconds, 25> recentConnectionTimings;
+        std::array<Seconds, 25>::iterator front;
```
Can we use an index instead so we can give it a default initializer and so it'll be a little more straightforward?

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -409,8 +409,10 @@ void NetworkResourceLoader::startNetworkLoad(ResourceRequest&& request, FirstLoa
     if (request.wasSchemeOptimisticallyUpgraded()) {
-        // FIXME: This timeout should be adaptive based on network conditions
-        request.setTimeoutInterval(10);
+        long optimisticUpgradeTimeout { 3 };
```
setTimeoutInterval takes a double.  I see no reason to only have an integer number of seconds.  Can we use a Seconds instead?

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -1082,6 +1084,9 @@ void NetworkResourceLoader::didFinishLoading(const NetworkLoadMetrics& networkLo
         m_connection->addNetworkLoadInformationMetrics(coreIdentifier(), networkLoadMetrics);
 
+    if (auto* networkSession = m_connection->networkSession())
+        networkSession->recordHTTPSConnectionTiming(networkLoadMetrics);
```
We could use the incomplete metrics when we receive a response to get it a little faster.

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkSession.h:
```diff
@@ -353,6 +356,13 @@ class NetworkSession : public WebCore::SWServerDelegate, public CanMakeCheckedPt
+        std::array<Seconds, 25> recentConnectionTimings { };
+        unsigned frontIndex { 0 };
+        bool isArrayFull { false };
+        double currentMovingAverage;
```
{ 0 }

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkSession.cpp:
```diff
@@ -714,6 +715,59 @@ void NetworkSession::setEmulatedConditions(std::optional<int64_t>&& bytesPerSeco
+    recentTiming.recentConnectionTimings[recentTiming.frontIndex++] = connectionEstablishmentTime;
+    ASSERT(recentTiming.frontIndex <= recentTiming.recentConnectionTimings.size());
+
+    if (recentTiming.frontIndex == recentTiming.recentConnectionTimings.size()) {
```
We could be conservative and say >= instead of ==

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkSession.h:
```diff
@@ -353,6 +356,13 @@ class NetworkSession : public WebCore::SWServerDelegate, public CanMakeCheckedPt
+    struct RecentHTTPSConnectionTiming {
+        std::array<Seconds, 25> recentConnectionTimings { };
+        unsigned frontIndex { 0 };
+        bool isArrayFull { false };
```
This is kind of an implementation of a Deque without allocation.  It might be worth using WTF::Deque instead.  One allocation won't hurt much, and it'll simplify the logic quite a bit.

sysrqb replied:
Good thought, that would simplify this.

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -878,6 +881,11 @@ void NetworkResourceLoader::didReceiveResponse(ResourceResponse&& receivedRespon
 
+    if (CheckedPtr networkSession = m_connection->networkSession()) {
+        if (auto metrics = computeResponseMetrics(m_response))
+            networkSession->recordHTTPSConnectionTiming(*metrics);
```
We aren't recording the connection timing of just HTTPS connections.  We probably want to rename this.

sysrqb replied:
Good catch, I must've dropped that at some point. I think we do want this scoped to new TLS connections.

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkSession.cpp:
```diff
@@ -714,6 +715,59 @@ void NetworkSession::setEmulatedConditions(std::optional<int64_t>&& bytesPerSeco
+    if (metrics.reusedConnection())
+        return;
+
+    Seconds connectionEstablishmentTime = metrics.connectEnd - metrics.fetchStart;
```
If there is a TLS handshake this will represent 2 roundtrips, but if not it will represent 1 roundtrip.  That will probably mess up the running average.

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkSession.cpp:
```diff
@@ -714,6 +715,48 @@ void NetworkSession::setEmulatedConditions(std::optional<int64_t>&& bytesPerSeco
+
+    double average = std::accumulate(connectionTimes.begin(), connectionTimes.end(), Seconds { }).seconds() / connectionTimes.size();
+    std::for_each(connectionTimes.begin(), connectionTimes.end(), [&average, alphaSmoothing](auto&& timing) {
+        average = alphaSmoothing*timing.seconds() + (1 - alphaSmoothing)*average;
```
spaces before and after the star.

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkSession.cpp:
```diff
@@ -714,6 +715,48 @@ void NetworkSession::setEmulatedConditions(std::optional<int64_t>&& bytesPerSeco
+    // Where x_0 is the average of all recent connection timings, and alpha is 0.75.
+
+    double average = std::accumulate(connectionTimes.begin(), connectionTimes.end(), Seconds { }).seconds() / connectionTimes.size();
+    std::for_each(connectionTimes.begin(), connectionTimes.end(), [&average, alphaSmoothing](auto&& timing) {
```
I'm not sure we get much benefit of using std::for_each over a c++11-style for loop.

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkSession.cpp:
```diff
@@ -714,6 +715,48 @@ void NetworkSession::setEmulatedConditions(std::optional<int64_t>&& bytesPerSeco
+    recentTiming.recentConnectionTimings.append(connectionEstablishmentTime);
+
+    auto newMovingAverage = connectionTimesMovingAverage(recentTiming.recentConnectionTimings);
+    newMovingAverage = std::max(minumumConnectionTimeout, newMovingAverage*computedTimeoutScalingFactor);
```
spaces before and after the star

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkSession.cpp:
```diff
@@ -714,6 +715,48 @@ void NetworkSession::setEmulatedConditions(std::optional<int64_t>&& bytesPerSeco
+
+    auto newMovingAverage = connectionTimesMovingAverage(recentTiming.recentConnectionTimings);
+    newMovingAverage = std::max(minumumConnectionTimeout, newMovingAverage*computedTimeoutScalingFactor);
+    if (newMovingAverage != recentTiming.currentMovingAverage) {
```
These will basically never be equal in practice.

sysrqb replied:
This will happen when we're repeatedly below the minimum timeout (3 seconds).

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkSession.h:
```diff
@@ -353,6 +357,12 @@ class NetworkSession : public WebCore::SWServerDelegate, public CanMakeCheckedPt
 
+    struct RecentHTTPSConnectionTiming {
+        static constexpr unsigned maxEntries { 25 };
+        Deque<Seconds, maxEntries> recentConnectionTimings { };
```
This is fine, but Deque has a default constructor so the { } don't do anything.

sysrqb committed:
Set timeout for optimistically upgraded navigations based on current network conditions
https://bugs.webkit.org/show_bug.cgi?id=278504
rdar://134458774

Reviewed by Alex Christensen.

This change implements a mechanism for dynamically setting the connection
timeout for requests that were optimistically upgraded. Currently, the timeout
is hard-coded, but there won't be a single hard-coded value that will be
appropriate for all networks. Therefore, in this patch I added a mechanism for
tracking the last 25 requests that used newly established TLS connections and
computes the exponentially weighed moving average (EWMA).

This patch sets a minimum (and default) timeout of 3 seconds, and the EMWA
skews the average toward more recent entries.

* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::startNetworkLoad):

Apply the appropriate timeout.

(WebKit::NetworkResourceLoader::didReceiveResponse):

Record this load, if appropriate.

* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::NetworkSession):
(WebKit::connectionTimesMovingAverage):
(WebKit::NetworkSession::recordHTTPSConnectionTiming):

Implement EWMA and compute the current average.

* Source/WebKit/NetworkProcess/NetworkSession.h:
(WebKit::NetworkSession::currentHTTPSConnectionAverageTiming const):

Canonical link: https://commits.webkit.org/283494@main
